doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,297 @@
+
+```
+query         rating    document      title_score view_count  recency  pagerank ...
+harry potter  0         Pottery barn  0.2         9876        0.12     9876     ...
+harry potter  3         Harry Potter  0.2         1234        1234     1234     ...
+...
+```
+
+For example, take the SVM dataset below for the query 'harry potter' (qid:1)
+
+```
+0 qid:1 popularity:0.3 title_score:0.2 description_score:0.3 recency:0.9 pagerank:5 ...
+3 qid:1 popularity:0.1 title_score:0.6 description_score:0.7 recency:0.5 pagerank:2 ...
+```
+
+These graded query:document pairs and their features can be used
+to train a model.
+
+The model might learn that title and description scores are slightly
+more useful that popularity and recency, and how much, in certain
+circumstances.
+
+Rather than manually tuning boosts, a model learns from relevancy
+judgements the role that features should play when providing relevant results.
+
+This is a high level overview of what Learning To Rank is for our use case.
+There has been a lot of research in this area, and within the fields of
+information retrieval & machine learning which Learning To Rank intersects.
+There is a lot more out there to learn!
+
+For more details about how we plan to implement Learning To Rank, see
+the implementation section below.
+
+## Motivation
+
+We have decided to try Learning To Rank because it has the potential
+to greatly improve search relevance in a way that is:
+
+- automatable
+- scalable
+- more maintainable
+- more successful
+
+**Automatable** because the model won't require human involvement to
+tune field boosts on the latest user data
+
+**Scalable** because automation; but also because the model should
+provide better results with more data.
+
+**Maintainable** because relevancy tuning can be automated to not become
+stale.
+
+**Successful** because results should achieve higher nDCG scores and
+click-through rates than we can achieve through manual tuning alone.
+
+### The problem
+
+Manual relevancy tuning can work, but doesn't scale well. Eventually
+the process becomes a game of whack-a-mole. If we downgrade popularity
+boosting, then this query might have better results, but other queries
+might have worse results.
+
+Hand-tuning relevance doesn't work very well because field boosts
+are complex, and difficult to maintain.
+
+Moreover field boosts might have dependencies: some document types might
+benefit from a stronger title boost than other document types, for
+instance. Expressing that just with Elasticsearch would be prohibitively
+complex, and be almost impossible to maintain.
+
+### Validating Learning To Rank
+
+We assumed that we would be able to get better results using Learning
+To Rank. To validate this assumption, we worked on a spike.
+
+In a [validation spike][] we found that we could build a model that provided
+a better nDCG score than our current search algorithm for the queries we
+have manually gathered relevancy judgements for.
```

Comment:
I believe we tested this on regular site search? Have we tested how LTR performs on specialist finders, or other specialist areas?

Reply:
We've just been trying this out on site search, though when we run an AB test we should investigate the impact it has on other finders.